caikit_nlp/modules/text_generation/text_generation_local.py

```diff
@@ -562,41 +571,43 @@ def _preprocess_function(
         max_source_length: int,
         max_target_length: int,
         shuffle: bool,
+        use_iterable_dataset: bool,
     ):
         """Pre-process each example to get it prepared for training."""
-
-        # TODO: We are using a default verbalizer which is strictly tied to
-        # source training record currently. We need to figure out a better
-        # way to make verbalizer optional for build_task_tokenize_function
-        (
-            tokenize_function,
-            requires_unwrapping,
-        ) = base_model.build_task_tokenize_function(
-            tokenizer, max_source_length, max_target_length, verbalizer="{{input}}"
-        )
-
-        # dataset = datasets.load_dataset("billsum", split="ca_test", streaming=True)
-        # train_test_dataset = dataset.train_test_split(test_size=0.2)
-        dataset = TransformersIterableDataset.from_generator(
-            get, gen_kwargs={"train_stream": train_stream}
-        )
-
+        if use_iterable_dataset:
+            # Generator based
+            log.debug("Loading data as an iterable dataset")
+            dataset = TransformersIterableDataset.from_generator(
+                get, gen_kwargs={"train_stream": train_stream}
+            )
+        else:
+            # Convert the train stream to an normal dataset in memory
+            log.debug("Loading data as a normal dataset")
+            # TODO: Optimize and clean this up!
+            inputs = []
+            outputs = []
+            if base_model.REQUIRES_TOKEN_UNWRAPPING:
+                for substream in train_stream:
+                    for data in substream:
+                        inputs.append(data.input)
+                        outputs.append(data.output)
+            else:
+                for data in train_stream:
```

Comment:
One of these is actually the bug in causal LM that I'm resolving now. For the other one, we can use list comprehensions, but I wrote the first pass to iterate over things once instead of 2 separate comprehensions 🤔

Reply:
Removed the loops